### PR TITLE
Treat not instantiated vms

### PR DESCRIPTION
**Pull request by jvmafra**
- Treat waiting queue to allocate VMs that are not instantiated in first moment
- Persistence of running vms and waiting vms

### CODE REVIEW

File: nova/scheduler/preemption_policy/__init__.py [hunk 1]
```diff
@@ -143,7 +148,7 @@ def deallocate_instance(self, host_name, spec_obj):
 
         self._update_runtime(spec_obj)
 
-        # self._write_scheduler_state()
+        self._write_scheduler_state()
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/__init__.py]: I think that update_state_info would be a better name for this method.

File: nova/scheduler/preemption_policy/__init__.py [hunk 2]
```diff
@@ -223,22 +228,67 @@ def _get_policy(self, conf):
...
+            
+        try:
+            with open(self._status_file, 'rb') as f:
+                status_dict = pickle.load(f)
+                return status_dict
+        except Exception as e:
+            print e.message
+            return None
+        
+    def _write_scheduler_state(self):
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/__init__.py]: Change the name as suggested before.

File: nova/scheduler/preemption_policy/__init__.py [hunk 3]
```diff
@@ -223,22 +228,67 @@ def _get_policy(self, conf):
...
+    def validate_waiting_instances(self, waiting):
+        
+        for vm in waiting:
+
+            uuid = vm.instance_uuid
+
+            try:
+                instance = self.nova_client.servers.get(uuid)
+                status = getattr(instance, 'status', 'unknown-name')
+                if status == "ACTIVE":
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/__init__.py]: What does it means? Is the instance running?

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 4]
```diff
@@ -216,13 +165,82 @@ def _schedule(self, context, spec_obj):
...
 
         # return if spec exists in waiting queue
-        return spec_obj in self.preemption_policy.waiting
+        is_created = spec_obj in self.preemption_policy.waiting
+
+        if not is_created:
+            spec_obj.scheduler_hints['submit_time'] = [datetime.datetime.now()]
+            print "New request"
+        else:
+            print "Try to reallocate VM"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: The else here should be It is a old request, right?

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 5]
```diff
@@ -216,13 +165,82 @@ def _schedule(self, context, spec_obj):
...
+                print "Id = %s" % running_vm.instance_uuid
+                slo = running_vm.get_scheduler_hint('slo')
+                print "Slo = %s" % slo
+
+        print "Waiting Vms:"
+        for waiting_vm in self.preemption_policy.waiting:
+            print "Id = %s" % waiting_vm.instance_uuid
+            print "Slo = %s" % waiting_vm.get_scheduler_hint('slo')
+
     def _is_created(self, spec_obj):
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Maybe the name "was_previously_created" could be a better name for this method.

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 6]
```diff
@@ -216,13 +165,82 @@ def _schedule(self, context, spec_obj):
...
+                                                            spec_obj)
+
+        LOG.debug("Weighed %(hosts)s", {'hosts': weighed_hosts})
+
+        # scheduler_host_subset_size = # of computes in nova.conf
+        host_subset_size = CONF.filter_scheduler.host_subset_size
+        if host_subset_size < len(weighed_hosts):
+            weighed_hosts = weighed_hosts[0:host_subset_size]
+
+        print "Weighed hosts:"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Join these prints.

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 7]
```diff
@@ -152,34 +141,7 @@ def _schedule(self, context, spec_obj):
...
-
-                    i += 1
-
-                weighed_hosts = weighed_hosts[0:i]        
-
-            chosen_host = random.choice(weighed_hosts)
+            chosen_host = self._choose_host(hosts, spec_obj)
 
             host_state = chosen_host.obj
             print "CHOSEN HOST:"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Join these prints

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 8]
```diff
@@ -216,13 +165,82 @@ def _schedule(self, context, spec_obj):
...
+            max_weight = weighed_hosts[0].weight
+            weighed_hosts = [weighed_host for weighed_host in weighed_hosts \
+                            if weighed_host.weight == max_weight]
+
+        return random.choice(weighed_hosts)
+
+    def _consume_resource(self, is_new, host_name, spec_obj, chosen_host):
+        if is_new:
+            print "Adding instance in running list of %s" % host_name
+            self.preemption_policy.create_instance(spec_obj, host_name)
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Put a comment here making clear what this method does. I think it is important to clarify that create_instance and allocate_instance do almost the same things.

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 9]
```diff
@@ -190,22 +152,9 @@ def _schedule(self, context, spec_obj):
...
-            if is_new:
-                print "Adding instance in running list of %s" % host_name
-                self.preemption_policy.create_instance(spec_obj, host_name)
-                chosen_host.obj.consume_from_request(spec_obj)
-                print "Finishing set host name for spec_obj"
-            else:
-                print "VM exists. Update it in lists..."
-                self.preemption_policy.allocate_instance(host_name, spec_obj)
-                print "Vm allocated"
+            self._consume_resource(is_new, host_name, spec_obj, chosen_host)
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: The name of the method could be in plural (resources).

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 10]
```diff
@@ -216,13 +165,82 @@ def _schedule(self, context, spec_obj):
...
+
+        if not is_created:
+            spec_obj.scheduler_hints['submit_time'] = [datetime.datetime.now()]
+            print "New request"
+        else:
+            print "Try to reallocate VM"
+
+        return is_created
+
+    def _is_valid_instance(self, context, instance_uuid):
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Add comment about the method. What is considered a valid instance?

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 11]
```diff
@@ -231,11 +249,22 @@ def try_to_allocate_waiting_queue(self, context):
 
         print "DESTROY EVENT...TRYING TO ALLOCATE WAITING QUEUE..."
         print "Size = %d" % len(self.preemption_policy.waiting)
+        print "Waiting Vms:"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Maybe you could separate the method printStructures in two: printWaintinVms and printRunningVMs. Then, here you could use one of these methods.

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 12]
```diff
@@ -244,35 +273,58 @@ def try_to_allocate_waiting_queue(self, context):
...
             if selected_hosts is not None and len(selected_hosts) > 0:  # host found
-                waiting_vm_id = waiting_vm.instance_uuid
 
-                print "VM %s can be reallocated" % waiting_vm_id
+                self._allocate_instance(waiting_vm_id, selected_hosts, actual_host, waiting_vm, context)
 
-                instance_ref = self.nova_client.servers.get(waiting_vm_id)
+    def _allocate_instance(self, waiting_vm_id, selected_hosts, actual_host, waiting_vm, context):
+        instance_ref = self.nova_client.servers.get(waiting_vm_id)
+        print "VM %s can be reallocated" % waiting_vm_id
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: I think the print could be "VM X is being allocated"

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 13]
```diff
@@ -244,35 +273,58 @@ def try_to_allocate_waiting_queue(self, context):
...
-                    print "No need migration"
+        else:
 
-                else:
-                    print "Need migration"
-                    instance_ref.live_migrate(host_name, block_migration=True)
-                    waiting_vm.scheduler_hints['host_name'] = [host_name]
-                    instance_ref = self.nova_client.servers.get(waiting_vm_id)
+            if host_name == actual_host:
+                print "No need migration"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Change to "Migration is not needed"

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 14]
```diff
@@ -244,35 +273,58 @@ def try_to_allocate_waiting_queue(self, context):
...
-                task_state = getattr(instance_ref, 'OS-EXT-STS:task_state', 'unknown-name')
-
-                # Unpause the VM only when it is paused
-                while status != "PAUSED" and task_state == "migrating":
-                    time.sleep(1)
-                    instance_ref = self.nova_client.servers.get(waiting_vm_id)
-                    status = getattr(instance_ref, 'status', 'unknown-name')
-                    task_state = getattr(instance_ref, 'OS-EXT-STS:task_state', 'unknown-name')
+            else:
+                print "Need migration"
```
Comment (giovannifs) [inline nova/scheduler/preemptive_filter_scheduler.py]: Change to "Migration is needed"

File: nova/conductor/manager.py [hunk 15]
```diff
@@ -860,17 +860,24 @@ def schedule_and_build_instances(self, context, build_requests,
                                      request_specs, image,
                                      admin_password, injected_files,
                                      requested_networks, block_device_mapping):
-        legacy_spec = request_specs[0].to_legacy_request_spec_dict()
-        # print legacy_spec informations about the request
+
         try:
-            hosts = self._schedule_instances(context, legacy_spec,
-                        request_specs[0].to_legacy_filter_properties_dict())
+            if len(request_specs) == 1 and request_specs[0].get_scheduler_hint('host_name') != '':
```
[the remark this answers is not in the document]
Comment (alessandroliafook) [reply nova/conductor/manager.py]: Add a comment: Checking whether the specification has been previously scheduled
ok?

File: nova/conductor/manager.py [hunk 16]
```diff
@@ -860,17 +860,24 @@ def schedule_and_build_instances(self, context, build_requests,
...
+                legacy_spec = request_specs[0].to_legacy_request_spec_dict()
+                hosts = self._schedule_instances(context, legacy_spec,
+                                                 request_specs[0].to_legacy_filter_properties_dict())
         except Exception as exc:
             LOG.exception(_LE('Failed to schedule instances'))
             print exc.message
-            self._bury_in_cell0(context, request_specs[0], exc,
-                                build_requests=build_requests)
-            print 'err'
+            if len(request_specs) == 1:
```
[the remark this answers is not in the document]
Comment (alessandroliafook) [reply nova/conductor/manager.py]: The system is implemented with this possibility but never occurs in ours test cases

File: nova/scheduler/manager.py [hunk 17]
```diff
@@ -133,11 +133,12 @@ def delete_instance_info(self, context, host_name, instance_uuid):
         """
 
         print "NEW DESTROY..."
-        self.driver.update_lists_after_destroy(host_name, instance_uuid)
+        release = self.driver.destroy_and_release_resource(host_name, instance_uuid)
```
[the remark this answers is not in the document]
Comment (alessandroliafook) [reply nova/scheduler/manager.py]: Can be.. I will change!

File: nova/scheduler/preemption_policy/__init__.py [hunk 18]
```diff
@@ -35,23 +37,26 @@ class PreemptionPolicy(object):
     def __init__(self):
 
         self._status_file = self._get_scheduler_state_file(CONF)
+        self.nova_client = self._get_novaclient(CONF)
+        print "Reading from DB..."
```
[the remark this answers is not in the document]
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/__init__.py]: I will put "Loading structures...", ok?

File: nova/scheduler/preemption_policy/__init__.py [hunk 19]
```diff
@@ -35,23 +37,26 @@ class PreemptionPolicy(object):
...
-        #     print "Loading last state"
-        #     self.running = db_file['vms_running']
-        #     self.waiting = db_file['vms_waiting']
-        # else:
-        print "Initializing empty lists"
-        # Dictionary with key/value equals hostname/spec_obj that indicates the local where the
-        # instance of spec_obj is allocated
-        self.running = {}
+        if db_file:
+            print "Loading last state"
```
[the remark this answers is not in the document]
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/__init__.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 1 above)
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/__init__.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 2 above)
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/__init__.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 3 above)
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/__init__.py]: Exactly, do i leave a comment?

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 7 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Ok

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 9 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Ok

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 6 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Done

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 5 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Done

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 4 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Exactly.. I will change the other message to "Try allocate new request" and the else message to "Try allocate old request".

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 10 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (giovannifs) [reply nova/scheduler/preemption_policy/__init__.py]: Adding VM

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 8 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Ok. I use the docstring of the method to do this!

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 11 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Done

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 12 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Agreed

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 13 above)
Comment (alessandroliafook) [reply nova/scheduler/preemptive_filter_scheduler.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 3 above)
Comment (giovannifs) [reply nova/scheduler/preemption_policy/__init__.py]: But if the instance is running, should we remove it from waiting, shouldn't?

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 4 above)
Comment (giovannifs) [reply nova/scheduler/preemptive_filter_scheduler.py]: ok

**Pull request merged**